faststreams/input_stream.nim

```diff
@@ -1,192 +1,298 @@
-  # TODO: the result should use `from mem` once it's supported
-  result.head = unsafeAddr mem[0]
-  result.bufferEnd = offset(result.head, mem.len)
-  result.bufferEndPos = mem.len
-  result.reader = reader
-  result.asyncReader = asyncReader
-
-proc memoryStream*(mem: openarray[byte]): ByteStreamVar = # TODO: mark the result with `from mem`
-  new result
-  result[] = ByteStream.init(mem)
-
-proc memoryStream*(str: string): ByteStreamVar = # TODO: mark the result with `from str`
-  new result
-  result[] = init(ByteStream, str.toOpenArrayByte(0, str.len - 1))
-
-proc init*(T: type BufferedStream,
-           reader = StreamReader(nil),
-           asyncReader = AsyncStreamReader(nil)): BufferedStream =
-  result.init result.buffer, reader, asyncReader
-
-proc endPos*(s: InputStreamVar): int =
-  # TODO This needs to use a VTable for `system.File` based streams
+    stream.bufferStart = head
+
+  InputStreamHandle(s: stream)
+
+proc memoryInput*(mem: openarray[byte]): InputStreamHandle =
+  let head = unsafeAddr mem[0]
+  InputStreamHandle(s: InputStream(
+    head: head,
+    bufferEnd: offset(head, mem.len),
+    bufferEndPos: mem.len))
+
+proc memoryInput*(str: string): InputStreamHandle =
+  memoryInput str.toOpenArrayByte(0, str.len - 1)
+
+# TODO: Is this used, should we deprecate it?
+proc endPos*(s: InputStream): int =
+  doAssert s.vtable == nil or s.vtable.getLenSync != nil
   return s.bufferEndPos
 
-proc syncRead(s: var ByteStream): bool =
-  let bytesRead = s.reader(s.bufferStart, s.bufferSize)
+# TODO The return type here could be Option[Natural] if Nim had
+# the Option[range] optimisation that will make it equvalent to `int`.
+proc len*(s: InputStream): int {.raises: [Defect, IOError].} =
+  if s.vtable == nil:
+    distance(s.head, s.bufferEnd)
+  elif s.vtable.getLenSync != nil:
+    s.vtable.getLenSync(s)
+  else:
+    lengthUnknown
+
+template len*(s: AsyncInputStream): int =
+  len InputStream(s)
+
+proc bufferMoreDataSync(s: InputStream): bool =
+  # Returns true if more data was successfully buffered
+  if s.vtable == nil or s.vtable.readSync == nil:
+    return false
+
+  let bytesRead = s.vtable.readSync(s, s.bufferStart, s.bufferSize)
   if bytesRead == 0:
-    s.reader = nil
-    return true
+    # TODO close the input device
+    s.vtable = nil
+    return false
   else:
     s.bufferEnd = offset(s.bufferStart, bytesRead)
     s.bufferEndPos += bytesRead
-    return false
+    return true
 
-proc ensureBytes*(s: var ByteStream, n: int): bool =
+proc bufferMoreDataAsync(s: AsyncInputStream): Future[bool] {.async.} =
+  # Returns true if more data was successfully buffered
+  return false
+
+proc readable*(s: InputStream): bool =
+  if s.head != s.bufferEnd:
+    true
+  else:
+    s.bufferMoreDataSync()
+
+template readable*(sp: AsyncInputStream): bool =
+  let s = sp
+  if s.head != s.bufferEnd:
+    true
+  else:
+    faststreamsAwait bufferMoreDataAsync(s)
```

Comment:
Well, I consider this one of the most important aspects of the API. It solves the performance issues you were pointing out in the LibP2P reviews (invoking synchronization on every byte). It's also relevant for the varint parsing discussion where even in the async case, my method of parsing varints will be more efficient than the current ones because of this API.

In general, my goal with the FastStream API is to allow you to write the stream processing code in a simple way as if you are handling a synchronous stream. From such a simple definition of a say JSON parser or a Snappy decoder I can automatically produce the async version of the same code. The `readable` calls are the only place where synchronization with the input device is required and it's triggered only when necessary.

Reply:
Another way to look at this is that this is one of the first APIs following the "enforced awaits" principle. Such APIs are arguably less error-prone than the `Future[T]` interface. 

From the perspective of the user, the `stream.reading` calls are "blocking" both in the sync and async scenarios in the sense that they introduce an unpredictable delay. The async version adds only some small additional hazards when you use heap objects that might be modified during the "blocking" code, but I don't think these will be common in practice because in a stream processor you are usually either transforming one form of data to another or you are populating a freshly created end-result object.

Reply:
> It solves the performance issues

read-in-chunks is fine - hiding `await` is not.

> additional hazards

I've solved several such bugs in NBC already - some that happened and some that were waiting to happen - I'm pretty sure there's more of them. This is real, and will get worse when it's hidden - until single-mutable-ownership happens, being explicit is the best we have - implicit capturing of `ref` in closures hides too much already and this would hide the out-of-order execution as well - not good.

> enforced awaits

this is orthogonal and having two "modes" is worse than having one - this library should not change the way chronos works unless it offers a complete replacement of the full chronos api, because otherwise, people that have "learned" to look for `await` will now be confused and vice versa. it's already a problem that people don't understand the differences between `waitfor/asynccheck/await` etc - introducing "one more" hidden way of blocking is not the way.

> Future[T] interface.

... is a corner-stone of chronos though in that you may want to add a timeouts etc - futures compose in the scheduler and in user code.

> heap objects

like for example like the libp2p long-lived peer that has a stream through which you read the data it sends - this is more common that you'd expect, specially in a networking application where the stream "trickles" instead of being read in one logical go.

Reply:
Try to implement in your head a consumer for something like JSON by reading chunks. You'll discover that it would be extremely complicated, because you'll have to write a lexer and a parser that can be pushed incomplete data. The JsonParser object will have to be a state machine that can be interrupted in any possible state (e.g. mid-way parsing a string). It would be extremely difficult to write correctly.

That's why I was pointing out that with the Chronos and LibP2P reforms you've discovered some of the tenets of the Pipelines design, but not all of them. Some protocols and formats are much easier to write when you write programming logic that pulls data out of the stream instead of being given the readMessage's `openArray` push mechanism.

I think the testament for this will be the simple code that I'll be able to demonstrate over and over again for any protocol or format that you throw at FastStreams. If you are able to demonstrate hazards and bugs in the code, we can talk about it.

Don't forget that this design is the evolution of 15 years of streaming APIs at Microsoft, not something you can easily dismiss on gut feelings.

Reply:
well, the chronos stuff solves a different problem: changing the object type that's being streamed from `byte` to `Something` - ie converting back and forth between types: TCP is a stream of frames, a blockchain is a stream of blocks - you'll agree that's not covered by faststreams?

Reply:
The challenge I gave you is exactly about converting a stream of bytes to a stream of strongly-typed values. JSON here serves just as an example - with any protocol or format based on a formal grammar, you'll face similar challenges. My claim is that with FastStreams API, I'll be able to implement any such format or protocol with relative ease, while the `readMessage` API will be pretty much useless and you'll end up implementing a buffer in the consumer that will have to be populated with the entire message before you can start parsing.